`r\managed_disk` Add support for `on_demand_bursting_enabled`

myc2h6o opened the pull request:
- Support disk bursting https://docs.microsoft.com/en-us/azure/virtual-machines/disk-bursting
- It can only be enabled for `Premium_LRS` and `Premium_ZRS`, with disk size greater than 512 GB
- Test result:
$ TF_ACC=1 go test -v ./internal/services/compute -run=withOnDemandBursting -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"                      === RUN   TestAccAzureRMManagedDisk_create_withOnDemandBurstingEnabled
=== PAUSE TestAccAzureRMManagedDisk_create_withOnDemandBurstingEnabled
=== RUN   TestAccAzureRMManagedDisk_update_withOnDemandBurstingEnabled
=== PAUSE TestAccAzureRMManagedDisk_update_withOnDemandBurstingEnabled
=== CONT  TestAccAzureRMManagedDisk_create_withOnDemandBurstingEnabled
=== CONT  TestAccAzureRMManagedDisk_update_withOnDemandBurstingEnabled
--- PASS: TestAccAzureRMManagedDisk_create_withOnDemandBurstingEnabled (346.92s)
--- PASS: TestAccAzureRMManagedDisk_update_withOnDemandBurstingEnabled (543.01s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/compute       543.700s

tombuildsstuff commented on internal/services/compute/managed_disk_resource.go:
```diff
@@ -356,6 +361,21 @@ func resourceManagedDiskCreate(d *pluginsdk.ResourceData, meta interface{}) erro
+		}
+
+		if diskSizeGB != 0 && diskSizeGB <= 512 {
+			return fmt.Errorf("`bursting_enabled` can only be set to true when `disk_size_gb` is larger than 512")
```
```suggestion
			return fmt.Errorf("`bursting_enabled` can only be set to true when `disk_size_gb` is larger than 512GB")
```

myc2h6o replied:
GB added

tombuildsstuff commented on internal/services/compute/managed_disk_resource.go:
```diff
@@ -514,6 +536,24 @@ func resourceManagedDiskUpdate(d *pluginsdk.ResourceData, meta interface{}) erro
+		}
+
+		if diskSizeGB != 0 && diskSizeGB <= 512 {
+			return fmt.Errorf("`bursting_enabled` can only be set to true when `disk_size_gb` is larger than 512")
```
```suggestion
			return fmt.Errorf("`bursting_enabled` can only be set to true when `disk_size_gb` is larger than 512GB")
```

myc2h6o replied:
GB added

tombuildsstuff commented on internal/services/compute/managed_disk_resource.go:
```diff
@@ -713,6 +753,7 @@ func resourceManagedDiskRead(d *pluginsdk.ResourceData, meta interface{}) error
 			}
 		}
 		d.Set("trusted_launch_enabled", trustedLaunchEnabled)
+		d.Set("bursting_enabled", props.BurstingEnabled)
```
the API doesn't necessarily return this for older resources - do we need to default this on our side:

```suggestion
		burstingEnabled := false
		if props.BurstingEnabled != nil {
			burstingEnabled = *props.BurstingEnabled
		}
		d.Set("bursting_enabled", burstingEnabled)
```

myc2h6o replied:
Thanks for pointing out, updated

tombuildsstuff commented on internal/services/compute/managed_disk_resource_test.go:
```diff
@@ -475,6 +475,28 @@ func TestAccAzureRMManagedDisk_create_withTrustedLaunchEnabled(t *testing.T) {
+
+	data.ResourceTest(t, r, []acceptance.TestStep{
+		{
+			Config: r.empty(data),
```
can we add a separate test which confirms the Create flow works too, preferably toggling that on/off in the update too

myc2h6o replied:
I added a test step create_withOnDemandBursting which explicitly set the property to false before upate_withOnDemandBursting so the update will toggle the setting from false to true.
The on-demand bursting can only be turned off after 12 hours after enabling it, so i only tested the off->on scenario

tombuildsstuff commented on internal/services/compute/managed_disk_resource.go:
```diff
@@ -356,6 +361,21 @@ func resourceManagedDiskCreate(d *pluginsdk.ResourceData, meta interface{}) erro
 
+	if d.Get("bursting_enabled").(bool) {
+		switch storageAccountType {
+		case string(compute.StorageAccountTypesPremiumLRS):
+		case string(compute.StorageAccountTypesPremiumZRS):
+		default:
+			return fmt.Errorf("`bursting_enabled` can only be set to true when `storage_account_type` is set to `Premium_LRS` or `Premium_ZRS`")
+		}
+
+		if diskSizeGB != 0 && diskSizeGB <= 512 {
```
the API appears to also support this for smaller Standard SSD's too:

> An on-demand bursting model, where the disk bursts whenever its needs exceed its current capacity. This model incurs additional charges anytime the disk bursts. On-demand bursting is only available for premium SSDs larger than 512 GiB.
> A credit-based model, where the disk will burst only if it has burst credits accumulated in its credit bucket. This model does not incur additional charges when the disk bursts. Credit-based bursting is only available for premium and standard SSDs 512 GiB and smaller.

https://docs.microsoft.com/en-us/azure/virtual-machines/disk-bursting#disk-level-bursting

As such I think this check could ultimately be removed?

myc2h6o replied:
The credit-based bursting is enabled by default on all eligible disks, including Standard and Premium. The property here only controls on-demand bursting. But seems like the name is confusing, I'm changing the property name to `on_demand_bursting_enabled`.
https://docs.microsoft.com/en-us/azure/virtual-machines/disks-enable-bursting
![image](https://user-images.githubusercontent.com/10579712/141427085-655707c9-f899-4490-8266-ef6b03c33054.png)

tombuildsstuff commented on internal/services/compute/managed_disk_resource.go:
```diff
@@ -514,6 +536,24 @@ func resourceManagedDiskUpdate(d *pluginsdk.ResourceData, meta interface{}) erro
 	}
 
+	if burstingEnabled {
+		switch storageAccountType {
+		case string(compute.StorageAccountTypesPremiumLRS):
+		case string(compute.StorageAccountTypesPremiumZRS):
+		default:
+			return fmt.Errorf("`bursting_enabled` can only be set to true when `storage_account_type` is set to `Premium_LRS` or `Premium_ZRS`")
+		}
+
+		if diskSizeGB != 0 && diskSizeGB <= 512 {
```
the API appears to also support this for smaller Standard SSD's too:

> An on-demand bursting model, where the disk bursts whenever its needs exceed its current capacity. This model incurs additional charges anytime the disk bursts. On-demand bursting is only available for premium SSDs larger than 512 GiB.
> A credit-based model, where the disk will burst only if it has burst credits accumulated in its credit bucket. This model does not incur additional charges when the disk bursts. Credit-based bursting is only available for premium and standard SSDs 512 GiB and smaller.

https://docs.microsoft.com/en-us/azure/virtual-machines/disk-bursting#disk-level-bursting

As such I think this check could ultimately be removed?

myc2h6o replied:
Same as the check in create

tombuildsstuff commented on website/docs/r/managed_disk.html.markdown:
```diff
@@ -133,6 +133,8 @@ The following arguments are supported:
 
 -> **Note:** Trusted Launch can only be enabled when `create_option` is `FromImage` or `Import`.
 
+* `bursting_enabled` (Optional) Specifies if the disk-level bursting is enabled for the Managed Disk. Defaults to `false`.
```
```suggestion
* `bursting_enabled` (Optional) Specifies if the disk-level bursting is enabled for the Managed Disk. Defaults to `false`.

-> **Note:** More information on [Credit and On-Demand Bursting can be found in the documentation](https://docs.microsoft.com/azure/virtual-machines/disk-bursting#disk-level-bursting).
```

myc2h6o replied:
Doc updated

tombuildsstuff reviewed (changes requested):
hi @myc2h6o 

Thanks for this PR.

Taking a look through the documentation here it appears that this is available both for Standard and Premium SSD's with different constraints - as such I believe we need to update the validation to account for this - but if we can fix those up then this should otherwise be good to go.

Thanks

myc2h6o commented:
Closed by mistake, reopened

myc2h6o commented:
Hi @tombuildsstuff thanks for reviewing the pr, I've updated the change according to your suggestions, and updated the property name to `on_demand_bursting_enabled` so that it will not be confused with credit-based bursting, please take another look

katbyte reviewed (approved):
Thanks @myc2h6o - LGTM 🎈
